crates/ra_hir/src/ty/tests.rs

```diff
@@ -2695,6 +2695,28 @@ fn main() {
     );
 }
 
+#[test]
+fn derive_macro_expanded() {
+    let (db, pos) = MockDatabase::with_position(
+        r#"
+//- /main.rs
+trait Clone {
+    fn clone(&self) -> Self;
+}
+
+#[derive(Clone)]
+struct S {}
+
+fn test() {
+    let s = S {};
+    s.clone()<|>;
+}
+"#,
+    );
+
+    assert_eq!("S", type_at_pos(&db, pos));
```

Comment:
@flodiebold @matklad I've implemented an inference test, but it doesn't seem to work. The type inferred is `{unknown}`. The problem is that `#[derive(Clone)]` expands into `impl Clone for S {}` token tree, but that doesn't seem to be correctly reflected further in HIR. If I replace derive with `impl` block, everything works fine. Also, if I produce an impl block with `macro_rules!` macro, the result is `{unknown}` too.

Reply:
I think the problem is that macros are expanding into different file or module. Or some plumbing is still missing.

Reply:
Ah, good to know... My guess would be that it's because impl blocks are collected separate from name resolution, though I don't know the macro infrastructure so well so I might be wrong. I'm not sure what the proper way to fix that would be.

Reply:
@matklad What do you think we should do here? Currently impl blocks are collected completely separately from nameres, but to keep that up we'd have to do macro expansion again there, I guess... Should `DefCollector` just collect impl blocks as well?

Reply:
Yeah, I kinda stuck with this issue. Should try to collect `impl` blocks too somehow.

Reply:
impl block collection was imled in https://github.com/rust-analyzer/rust-analyzer/pull/1861

Reply:
Now, this test passes! 🎉